### ISSUE TITLE
Durations not round-tripping through Crux

### ISSUE THREAD

**Issue by jarohen** [xtdb/xtdb#1226]
Thanks @dvingo for the repro:

https://github.com/dvingo/crux_1_12_data_literals_bug

1.10.0 is last known good, 1.11.0 is first known bad

```clojure
(crux/submit-tx crux-node
                [[:crux.tx/put {:crux.db/id "hi"
                                :my/duration (java.time.Duration/parse "PT10M")}]])

(crux/entity (crux/db crux-node) "hi")
;; => {:crux.db/id "hi", :my/duration (. java.time.Duration parse "PT10M")}
```

**Comment by jarohen** [comment]
We upgraded Nippy in 1.11.0 to 2.15.1 for ptaoussanis/nippy#130 - it seems that the encoding of `Duration`s changes in that upgrade. Before, it uses the `Serializable` implementation; after, it seems to use something akin to `print-dup` (hence the `(. java.time.Duration parse "PT10M")`).

[edit: it uses `pr-str`, but `print-method` in this case is being overridden by something that looks more like a `print-dup` implementation]

I'll see whether there's anything we can do our side to make this less of a gotcha - in the meantime, a workaround is to add `java.time.Duration` to Nippy's ['serializable whitelist'](https://cljdoc.org/d/com.taoensso/nippy/2.15.1/api/taoensso.nippy#*serializable-whitelist*), at which point the old behaviour seems to be restored.

**Comment by dvingo** [comment]
For anyone storing java.time objects in crux, this gets things working as of 2020-11-04:
```clojure
(taoensso.nippy/swap-serializable-whitelist!
     (fn [old]
       (conj old "java.time.*")))
```
For future reference, if nippy is updated, the way to specify allowed serializable classes will change:
https://github.com/ptaoussanis/nippy/blob/a98b0d8bb1eab52b8c488fd440025ffd750389fa/src/taoensso/nippy.clj#L389

**Comment by jarohen** [comment]
Alright, so:

tl;dr can repro this without Crux, so I'll be closing it as a Crux issue (given we have a workaround)

```clojure 
(-> (nippy/freeze (java.time.Duration/parse "PT10M")) 
    (nippy/thaw))
;; => (. java.time.Duration parse "PT10M")
```

Repro requires Nippy (I'm using 2.15.1 but looking at the changes in 3.x I'd guess it's the same there) and [time-literals](https://github.com/henryw374/time-literals) (particularly, its `read_write.cljc`, enabled by default in this case by [tick](https://github.com/juxt/tick))

Nippy alone fails to serialize a Duration in 2.15.1, because it's not on the 'serializable-whitelist' by default - it looks like `java.time.*` has since been added but commented out, with the comment 'safe?' - I trust that @ptaoussanis hasn't yet included this for good reason, but good that it's being considered :)

It then falls back to `write-readable`, essentially Encore's pr-edn/read-edn. `time-literals` overrides the `print-method` for `java.time.Duration` to the above string, which is then frozen/thawed in Nippy. I'd guess this `print-method` works fine in code when you're evaluating the result, but in EDN you don't - an issue raised at https://github.com/henryw374/time-literals/issues/3 

James

**Comment by ptaoussanis** [comment]
Hi folks,

Just pushed https://github.com/ptaoussanis/nippy/releases/tag/v3.1.0 to Clojars which:

- Adds native thaw/freeze for `java.time`: `Instant`, `Duration`, `Period`.
- Whitelists a number of other common `java.time` classes for `Serializable`.
- Adds a deprecated `swap-serializable-whitelist!` util to help prevent breakages during migration.

**Comment by jarohen** [comment]
Cheers @ptaoussanis! :pray:
